### PR TITLE
clean now runs after all SpotlessTask.  Fixes #159.

### LINKED ISSUE #159: Error from spotless after running a clean before ktlint linting
Running a `./gradlew clean spotlessCheck` or `./gradlew clean spotlessApply` sometimes results in an error like this:

```
Step 'ktlint' found problem in 'src/main/kotlin/com/plexxi/fitting/api/ComputationalInterfaces.kt':
null
java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.diffplug.spotless.kotlin.KtLintStep$State.createFormat(KtLintStep.java:85)
        at com.diffplug.spotless.FormatterStepImpl$Standard.format(FormatterStepImpl.java:76)
        at com.diffplug.spotless.FormatterStep$Strict.format(FormatterStep.java:76)
        at com.diffplug.spotless.Formatter.compute(Formatter.java:230)
        at com.diffplug.spotless.Formatter.isClean(Formatter.java:167)
        at com.diffplug.gradle.spotless.SpotlessTask.check(SpotlessTask.java:216)
        at com.diffplug.gradle.spotless.SpotlessTask.performAction(SpotlessTask.java:172)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$IncrementalTaskAction.doExecute(DefaultTaskClassInfoStore.java:179)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:135)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:122)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:121)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:97)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:87)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:625)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:580)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NoClassDefFoundError: com/github/shyiko/ktlint/core/Rule
        at com.github.shyiko.ktlint.ruleset.standard.StandardRuleSetProvider.get(StandardRuleSetProvider.kt:8)
        ... 57 more
Caused by: java.lang.ClassNotFoundException: com.github.shyiko.ktlint.core.Rule
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 58 more
```

I opened this [issue with ktlint](https://github.com/shyiko/ktlint/issues/95) to which I got the response:
> `java.lang.ClassNotFoundException: com.github.shyiko.ktlint.core.Rule` indicates that `com.github.shyiko.ktlint:ktlint-core` (transitive dependency of `com.github.shyiko:ktlint`) isn't available on the classpath. Looks like artifact resolution / class loading / misconfiguration issue (most definitely not a `ktlint` problem).

I'm not sure why this happens only immediately after running a `clean`.

## Version Info
```
------------------------------------------------------------
Gradle 4.2
------------------------------------------------------------

Build time:   2017-09-20 14:48:23 UTC
Revision:     5ba503cc17748671c83ce35d7da1cffd6e24dfbd

Groovy:       2.4.11
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_92 (Oracle Corporation 25.92-b14)
OS:           Mac OS X 10.12.6 x86_64
```

Spotless Version: `3.5.1`
Ktlint Version: `0.9.2`

### PR DESCRIPTION
**Pull request by nedtwigg**

### CODE REVIEW

File: plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java [hunk 1]
```diff
@@ -107,10 +107,12 @@ public Object doCall(TaskExecutionGraph graph) {
...
 
-		// clear spotless' cache when the user does a clean
+		// clear spotless' cache when the user does a clean, but only after any spotless tasks
+		Task clean = project.getTasks().getByName(BasePlugin.CLEAN_TASK_NAME);
+		clean.doLast(unused -> SpotlessCache.clear());
 		project.getTasks()
-				.matching(task -> task.getName().equals("clean"))
-				.all(task -> task.doLast(unused -> SpotlessCache.clear()));
+				.matching(task -> task instanceof SpotlessTask)
+				.all(task -> clean.mustRunAfter(task));
```
Comment (jbduncan) [inline plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java]: @nedtwigg IIUC from what you said at https://github.com/diffplug/spotless/issues/159#issuecomment-339445006, this line ought to be, `.all(task -> task.mustRunAfter(clean))`. Have I misunderstood something here?

Comment (jbduncan) [reply plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java]: It's also unclear to me if `clean.mustRunAfter(task)` forces `clean` to run regardless of whether or not a user requested it on top of `spotless{Apply,Check}`.

In other words, do `./gradlew spotlessCheck` and `./gradlew {clean} spotlessCheck {clean}` do the same thing now?

Comment (nedtwigg) [reply plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java]: `mustRunAfter` is [not the same as `dependsOn`](https://docs.gradle.org/current/javadoc/org/gradle/api/Task.html#setMustRunAfter(java.lang.Iterable))

> this line ought to be, `.all(task -> task.mustRunAfter(clean))`

Good catch!  When I wrote this code, I meant for the order to be `spotless` then `clean`, but that's silly.  "Run the compiler, then clean the output".  lol.  I'll push a fix.

File: plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java [hunk 2]
```diff
@@ -107,10 +107,12 @@ public Object doCall(TaskExecutionGraph graph) {
...
 
-		// clear spotless' cache when the user does a clean
+		// clear spotless' cache when the user does a clean, but only after any spotless tasks
+		Task clean = project.getTasks().getByName(BasePlugin.CLEAN_TASK_NAME);
+		clean.doLast(unused -> SpotlessCache.clear());
 		project.getTasks()
-				.matching(task -> task.getName().equals("clean"))
-				.all(task -> task.doLast(unused -> SpotlessCache.clear()));
+				.matching(task -> task instanceof SpotlessTask)
+				.all(task -> task.mustRunAfter(clean));
```
Comment (JLLeitschuh) [inline plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java]: Gradle has a method (at least in the newer versions) that's something like:
```java
project.getTasks().withType(SpotlessTask.class).all(task -> task.mustRunAfter(clean));
```

Comment (nedtwigg) [reply plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java]: Awesome, thanks!

File: plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java (hunk 1 above)
Comment (jbduncan) [reply plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java]: > `mustRunAfter` is not the same as `dependsOn`

Cool! Thanks for the clarification.

> Good catch!

You're very welcome!

> When I wrote this code, I meant for the order to be spotless then clean, but that's silly. "Run the compiler, then clean the output". lol. I'll push a fix.

haha, SGTM. :)

**Comment by nedtwigg** [comment]
Unless there are objections, I'm gonna merge in an hour, so that @JLLeitschuh can do an integration test on `3.7.0-SNAPSHOT` and make sure #159 is solved for him.

**Comment by nedtwigg** [comment]
I agree a test would be nice, but I couldn't figure a way to reliably repeat the race condition.

**Pull request merged**